### ISSUE TITLE
InvalidOperationException by ODataUriParser

### ISSUE THREAD

**Issue by JefWight** [OData/odata.net#220]
When parsing, `UriUtils.TryUriStringToDateTimeOffset` catches `FormatException`, but `PlatformHelper.ConvertStringToDateTimeOffset` (in `ValidateTimeZoneInformationInDateTimeOffsetString`) throws `InvalidOperationException` instead.  The net result is that a call to `ODataUriParser.ParseUri` can result in an `InvalidOperationException` instead of an `ODataException`.

Not sure if the right fix is to `catch (InvalidOperationException)` or to `throw new FormatException`.

Call Stack:
```
System.InvalidOperationException: The time zone information is missing on the DateTimeOffset value '2015-05-31T19:24:52'. A DateTimeOffset value must contain the time zone information.
   at Microsoft.OData.Core.PlatformHelper.ValidateTimeZoneInformationInDateTimeOffsetString(String text)
   at Microsoft.OData.Core.UriParser.UriUtils.TryUriStringToDateTimeOffset(String text, DateTimeOffset& targetValue)
   at Microsoft.OData.Core.UriParser.ExpressionLexer.ParseFromDigit()
   at Microsoft.OData.Core.UriParser.ExpressionLexer.NextTokenImplementation(Exception& error)
   at Microsoft.OData.Core.UriParser.Parsers.UriQueryExpressionParser.ParseComparison()
   at Microsoft.OData.Core.UriParser.Parsers.UriQueryExpressionParser.ParseLogicalAnd()
   at Microsoft.OData.Core.UriParser.Parsers.UriQueryExpressionParser.ParseLogicalOr()
   at Microsoft.OData.Core.UriParser.Parsers.UriQueryExpressionParser.ParseExpressionText(String expressionText)
   at Microsoft.OData.Core.UriParser.ODataQueryOptionParser.ParseFilterImplementation(String filter, ODataUriParserConfiguration configuration, IEdmType elementType, IEdmNavigationSource navigationSource)
   at Microsoft.OData.Core.UriParser.ODataQueryOptionParser.ParseFilter()
   at Microsoft.OData.Core.UriParser.ODataUriParser.ParseUri()
```

**Comment by MrTomWhite** [comment]
It feels like the error message is explanatory as it is. 
"System.InvalidOperationException: The time zone information is missing on the DateTimeOffset value '2015-05-31T19:24:52'. A DateTimeOffset value must contain the time zone information."

You're making an invalid operation, so I can buy the exception it throws is correct. The message contains your given value and hints at a clear way to fix it. What more do we need here? 
-Tom White 

Date: Mon, 8 Jun 2015 12:43:03 -0700

**Comment by JefWight** [comment]
The error message is acceptable.  For most errors with parsing a URI, the core library properly catches things like FormatException and wraps it in an ODataException.  This allows me to catch those exceptions, knowing it's a client error, and respond with a 400 level response with a known format / message.  The ask is to make ODataUriParser.ParseUri to ultimately return an ODataException here, not an InvalidOperationException, making it behave like other client errors.

**Comment by karataliu** [comment]
I'd suggest that by changing the exception from InvalidOperationException to FormatException makes the API looks like more uniform, also we may consider missing timezone in string as a format exception. Besides, it doesn't quite make sense to throw exception inside a method named TryXX (Microsoft.OData.Core.UriParser.UriUtils.TryUriStringToDateTimeOffset).

But after that we would get following exception from ODataUriParser:
```
Syntax error at position 7 in '2015-05-31T19:24:32 ...'
```
This is because TryUriStringToDateTimeOffset failed (due to no timezone specified), and the parser would try its best to parse the literal into a reasonable value. Then it parses '2015' as a number, and fails on '-03' at position 7. 

In this case user might find it hard to figure out what happens here, as the string seems to be a valid DataTimeOffset literal (only timezone missing), and there's no additional hint for fixing it.

In conclusion, I think it makes sense to change the exception type here, and in future we may make some change to make inner format exception information available to user (maybe by aggregating them as an inner exception on the ODataException).

**Comment by chinadragon0515** [comment]
Same issue as #434

**Comment by chinadragon0515** [comment]
Code is merged, close it.
